### PR TITLE
feat: add customizable upgrade message on client/server version mismatch

### LINKED ISSUE #10718: feat: customize install message in cli
## Overview
Currently, when a user's `coder` CLI version is behind the control plane's version, they're prompted to update [here](https://github.com/coder/coder/blob/e7fd2cb1a64040e58ed26b0e45ac258311c58b52/cli/root.go#L819). For deployments with a custom install script, this message's link pointing to [coder.com](https://coder.com/) is not useful. A customer has requested that this URL (or the entire message) be customizable so it may point to their custom install script. 

## Impact
This would complete the use case of defining a custom installation routine in any deployment.

## Implementation details
The customer simply wants the URL to direct to their control plane rather than coder.com, but it would be best if this message could be fully customizable by users.

### PR DESCRIPTION
**Pull request by sreya**
fixes https://github.com/coder/coder/issues/10718

Screenshot 2024-01-11 at 6 31 18 PM

### CODE REVIEW

File: cli/testdata/server-config.yaml.golden [hunk 1]
```diff
@@ -433,6 +433,11 @@ client:
   # incorrectly can break SSH to your deployment, use cautiously.
   # (default: <unset>, type: string-array)
   sshConfigOptions: []
+  # The upgrade message to display to users when a client/server mismatch is
+  # detected. By default it instructs users to update using 'curl -L
+  # https://coder.com/install.sh | sh'.
+  # (default: <unset>, type: string)
+  cliUpgradeMessage: ""
```
Comment (kylecarbs) [inline cli/testdata/server-config.yaml.golden:440]: Why don't we default to `curl -L` in here? We have to make the request anyways.

File: cli/root.go [hunk 2]
```diff
@@ -824,11 +829,15 @@ func (r *RootCmd) checkVersions(i *clibase.Invocation, client *codersdk.Client)
 	if isConnectionError(err) {
 		return nil
 	}
-
 	if err != nil {
 		return xerrors.Errorf("build info: %w", err)
 	}
 
+	dconfig, err := client.UnprivilegedDeploymentConfig(ctx)
```
Comment (kylecarbs) [inline cli/root.go:836]: IMO you should toss the upgrade message in `buildinfo`. It's somewhat jank, but I think less so than making a new request and endpoint.

File: cli/root.go [hunk 3]
```diff
@@ -812,7 +812,12 @@ func formatExamples(examples ...example) string {
 	return sb.String()
 }
 
-func (r *RootCmd) checkVersions(i *clibase.Invocation, client *codersdk.Client) error {
+// checkVersions checks to see if there's a version mismatch between the client
+// and server and prints a message nudging the user to upgrade if a mismatch
+// is detected. forceCheck is a test flag and should always be false in production.
+//
+//nolint:revive
+func (r *RootCmd) checkVersions(i *clibase.Invocation, client *codersdk.Client, forceCheck bool) error {
```
Comment (Emyrk) [inline cli/root.go:820]: I think we can solve this without a `forceCheck` bool, will code something here and post

Comment (Emyrk) [reply cli/root.go:820]: Ahh it is because of the `devel` version that always returns a match :thinking:

File: cli/root.go [hunk 4]
```diff
@@ -826,24 +831,17 @@ func (r *RootCmd) checkVersions(i *clibase.Invocation, client *codersdk.Client)
...
+	fmtWarn := pretty.Sprint(cliui.DefaultStyles.Warn, fmtWarningText)
+	warning := fmt.Sprintf(fmtWarn, info.Version, strings.TrimPrefix(info.CanonicalVersion(), "v"), info.UpgradeMessage)
 
-	if !buildinfo.VersionsMatch(clientVersion, info.Version) {
-		warn := cliui.DefaultStyles.Warn
-		_, _ = fmt.Fprintf(i.Stderr, pretty.Sprint(warn, fmtWarningText), clientVersion, info.Version, strings.TrimPrefix(info.CanonicalVersion(), "v"))
+	if !buildinfo.VersionsMatch(clientVersion, info.Version) || forceCheck {
+		_, _ = fmt.Fprint(i.Stderr, warning)
 		_, _ = fmt.Fprintln(i.Stderr)
 	}
```
Comment (Emyrk) [inline cli/root.go:846]: This text is only printed in the `if` body, so put the formatting there too:

```golang
	if !buildinfo.VersionsMatch(expected, info.Version) {
		fmtWarningText := "version mismatch: client %s, server %s\n%s"

		fmtWarn := pretty.Sprint(cliui.DefaultStyles.Warn, fmtWarningText)
		warning := fmt.Sprintf(fmtWarn, info.Version, strings.TrimPrefix(info.CanonicalVersion(), "v"), info.UpgradeMessage)

		_, _ = fmt.Fprint(i.Stderr, warning)
		_, _ = fmt.Fprintln(i.Stderr)
	}
```

**Review by Emyrk — CHANGES_REQUESTED** [review]

File: codersdk/deployment.go [hunk 5]
```diff
@@ -2287,3 +2305,13 @@ func (c *Client) SSHConfiguration(ctx context.Context) (SSHConfigResponse, error
...
+
+func defaultUpgradeMessage() string {
+	// Our installation script doesn't work on Windows, so instead we direct the user
+	// to the GitHub release page to download the latest installer.
+	version := buildinfo.Version()
+	if runtime.GOOS == "windows" {
+		return fmt.Sprintf("download the server version from: https://github.com/coder/coder/releases/v%s", version)
+	}
+	return fmt.Sprintf("download the server version with: 'curl -L https://coder.com/install.sh | sh -s -- --version %s'", version)
+}
```
Comment (Emyrk) [inline codersdk/deployment.go:2317]: I think my comment got removed, reposting it (might see this twice??)

This is now running on the server, not the client. So the OS is likely to never be windows, and windows clients will get the linux/mac msg.

I think having the client send it's OS is "eh", so maybe just make the default an empty string. On the client side, if the server sends over an empty string, then use this function. Pushing the caller of this back to the client.

**Review by Emyrk — CHANGES_REQUESTED** [review]

File: codersdk/deployment.go (hunk 5 above)
Comment (sreya) [reply codersdk/deployment.go:2317]: :facepalm:

File: cli/root.go (hunk 3 above)
Comment (sreya) [reply cli/root.go]: Agreed definitely better

**Review by Emyrk — APPROVED** [review]

**Pull request merged**